File: apps/web/lib/api/conditions.ts
// Self-contained API bindings for the stage-10 Disbursement Condition ledger
// screen ("Điều kiện giải ngân"). Mirrors lib/api/client.ts conventions but
// shares no mutable state with it.
//
// Backend truth mirrored here: services/api/src/creditops/api/conditions.py and
// services/.../domain/conditions.py.
//   GET  /api/v1/cases/{caseId}/conditions                        -> Conditions (list + confirmable)
//   POST /api/v1/cases/{caseId}/conditions                        -> 201 Condition (bound to permitting decision)
//   POST /api/v1/cases/{caseId}/conditions/{conditionId}/transition -> Condition (validated edge)
//   POST /api/v1/cases/{caseId}/conditions/confirm                -> Confirmation (HG_DISBURSEMENT_CONDITIONS_CONFIRMED)
//
// A human-only ledger: agents never write a status. Verification / waiver /
// not-applicable are independent-checker acts; a waiver or not-applicable ruling
// requires an authority rationale. Confirmation is fail-closed and never
// satisfied by an empty ledger.

import { ApiClientError } from "./client";

const BFF_BASE_URL = "/api/creditops";
const CSRF_COOKIE_NAME = "__Host-creditops-csrf";
const CSRF_HEADER_NAME = "x-creditops-csrf";

// --- Domain enums (mirror services/.../domain/conditions.py) ---

export type GateStatus = "OPEN" | "SATISFIED";

export type ConditionStatus =
  | "PENDING"
  | "EVIDENCE_SUBMITTED"
  | "VERIFIED"
  | "FAILED"
  | "WAIVER_REQUESTED"
  | "WAIVED_BY_HUMAN"
  | "SUPERSEDED"
  | "NOT_APPLICABLE_BY_HUMAN";

// The exact 8-status closed set, in display order.
export const CONDITION_STATUSES: readonly ConditionStatus[] = [
  "PENDING",
  "EVIDENCE_SUBMITTED",
  "VERIFIED",
  "FAILED",
  "WAIVER_REQUESTED",
  "WAIVED_BY_HUMAN",
  "SUPERSEDED",
  "NOT_APPLICABLE_BY_HUMAN",
];

// The deterministic transition map (mirrors ALLOWED_TRANSITIONS). The UI derives
// each condition's target choices from its current status, so it never offers a
// forbidden edge; the server re-validates regardless.
export const ALLOWED_TRANSITIONS: Record<ConditionStatus, readonly ConditionStatus[]> = {
  PENDING: ["EVIDENCE_SUBMITTED", "WAIVER_REQUESTED", "NOT_APPLICABLE_BY_HUMAN", "SUPERSEDED"],
  EVIDENCE_SUBMITTED: ["VERIFIED", "FAILED", "SUPERSEDED"],
  FAILED: ["EVIDENCE_SUBMITTED", "WAIVER_REQUESTED", "SUPERSEDED"],
  WAIVER_REQUESTED: ["WAIVED_BY_HUMAN", "FAILED", "SUPERSEDED"],
  VERIFIED: ["SUPERSEDED"],
  WAIVED_BY_HUMAN: ["SUPERSEDED"],
  NOT_APPLICABLE_BY_HUMAN: ["SUPERSEDED"],
  SUPERSEDED: [],
};

// Targets reserved to the independent OPS checker authority (mirrors
// CHECKER_AUTHORITY_TARGETS): verification and the human waiver / not-applicable
// rulings. Every other move is an ordinary ops-officer workflow step.
export const CHECKER_AUTHORITY_TARGETS: readonly ConditionStatus[] = [
  "VERIFIED",
  "WAIVED_BY_HUMAN",
  "NOT_APPLICABLE_BY_HUMAN",
];

// Targets whose ruling IS the authority record and therefore REQUIRE a rationale
// (mirrors RATIONALE_REQUIRED_TARGETS).
export const RATIONALE_REQUIRED_TARGETS: readonly ConditionStatus[] = [
  "WAIVED_BY_HUMAN",
  "NOT_APPLICABLE_BY_HUMAN",
];

// The satisfied terminals a ledger may be confirmed on (mirrors
// CONFIRMABLE_STATUSES).
export const CONFIRMABLE_STATUSES: readonly ConditionStatus[] = [
  "VERIFIED",
  "WAIVED_BY_HUMAN",
  "NOT_APPLICABLE_BY_HUMAN",
];

// --- Response shapes ---

export interface DisbursementCondition {
  id: string;
  caseId: string;
  caseVersion: number;
  decisionId: string;
  conditionText: string;
  owner: string | null;
  dueDate: string | null;
  status: ConditionStatus | string;
  evidenceRefs: string[];
  createdAt: string;
}

export interface ConditionLedger {
  conditions: DisbursementCondition[];
  caseVersion: number;
  confirmable: boolean;
}

export interface ConditionConfirmation {
  gateType: string;
  status: GateStatus | string;
  caseVersion: number;
  dispositionRef: string;
}

export interface CreateConditionInput {
  conditionText: string;
  owner?: string;
  dueDate?: string;
}

export interface TransitionConditionInput {
  toStatus: ConditionStatus;
  rationale?: string;
  evidenceRefs?: string[];
}

// --- Defensive parsing ---

function asRecord(value: unknown): Record<string, unknown> {
  return typeof value === "object" && value !== null ? (value as Record<string, unknown>) : {};
}

function str(value: unknown): string {
  return typeof value === "string" ? value : value == null ? "" : String(value);
}

function strOrNull(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

function num(value: unknown): number {
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}

function bool(value: unknown): boolean {
  return value === true;
}

function strArray(value: unknown): string[] {
  return Array.isArray(value) ? value.map(str) : [];
}

function parseCondition(value: unknown): DisbursementCondition {
  const raw = asRecord(value);
  return {
    id: str(raw.id),
    caseId: str(raw.caseId),
    caseVersion: num(raw.caseVersion),
    decisionId: str(raw.decisionId),
    conditionText: str(raw.conditionText),
    owner: strOrNull(raw.owner),
    dueDate: strOrNull(raw.dueDate),
    status: str(raw.status),
    evidenceRefs: strArray(raw.evidenceRefs),
    createdAt: str(raw.createdAt),
  };
}

export function parseConditionLedger(value: unknown): ConditionLedger {
  const raw = asRecord(value);
  return {
    conditions: Array.isArray(raw.conditions) ? raw.conditions.map(parseCondition) : [],
    caseVersion: num(raw.caseVersion),
    confirmable: bool(raw.confirmable),
  };
}

export function parseDisbursementCondition(value: unknown): DisbursementCondition {
  return parseCondition(value);
}

export function parseConditionConfirmation(value: unknown): ConditionConfirmation {
  const raw = asRecord(value);
  return {
    gateType: str(raw.gateType),
    status: str(raw.status),
    caseVersion: num(raw.caseVersion),
    dispositionRef: str(raw.dispositionRef),
  };
}

export function allowedConditionTransitions(status: string): readonly ConditionStatus[] {
  return ALLOWED_TRANSITIONS[status as ConditionStatus] ?? [];
}

export function isRationaleRequired(target: ConditionStatus): boolean {
  return RATIONALE_REQUIRED_TARGETS.includes(target);
}

// --- Client ---

type Fetcher = typeof fetch;
type CsrfTokenProvider = () => string | null;

function readBrowserCsrfToken(): string | null {
  if (typeof document === "undefined") return null;
  for (const part of document.cookie.split(";")) {
    const index = part.indexOf("=");
    if (index < 0 || part.slice(0, index).trim() !== CSRF_COOKIE_NAME) continue;
    try {
      return decodeURIComponent(part.slice(index + 1).trim());
    } catch {
      return null;
    }
  }
  return null;
}

async function parseJson(response: Response): Promise<unknown> {
  const contentType = response.headers.get("content-type") ?? "";
  if (!contentType.includes("application/json")) return null;
  try {
    return await response.json();
  } catch {
    return null;
  }
}

function parseApiError(
  body: unknown,
): { code: string; messageVi: string; retryable: boolean } | null {
  if (typeof body !== "object" || body === null) return null;
  const raw = body as Record<string, unknown>;
  if (typeof raw.code !== "string") return null;
  return {
    code: raw.code,
    messageVi: typeof raw.messageVi === "string" ? raw.messageVi : "",
    retryable: typeof raw.retryable === "boolean" ? raw.retryable : false,
  };
}

export class ConditionsApiClient {
  private readonly baseUrl: string;

  constructor(
    baseUrl = BFF_BASE_URL,
    private readonly fetcher: Fetcher = (input, init) => fetch(input, init),
    private readonly csrfTokenProvider: CsrfTokenProvider = readBrowserCsrfToken,
  ) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
  }

  async getLedger(caseId: string): Promise<ConditionLedger> {
    return parseConditionLedger(
      await this.request(`/api/v1/cases/${encodeURIComponent(caseId)}/conditions`),
    );
  }

  // Opens ONE condition bound to a permitting decision. 409
  // CONDITIONS_REQUIRE_APPROVAL_DECISION when no approval exists yet.
  async createCondition(
    caseId: string,
    input: CreateConditionInput,
  ): Promise<DisbursementCondition> {
    return parseDisbursementCondition(
      await this.request(`/api/v1/cases/${encodeURIComponent(caseId)}/conditions`, {
        method: "POST",
        body: JSON.stringify(compact(input)),
      }),
    );
  }

  // Moves ONE condition along a validated edge under the correct authority. 422
  // RATIONALE_REQUIRED for a waiver / not-applicable ruling with no rationale.
  async transition(
    caseId: string,
    conditionId: string,
    input: TransitionConditionInput,
  ): Promise<DisbursementCondition> {
    return parseDisbursementCondition(
      await this.request(
        `/api/v1/cases/${encodeURIComponent(caseId)}/conditions/${encodeURIComponent(
          conditionId,
        )}/transition`,
        { method: "POST", body: JSON.stringify(compact(input)) },
      ),
    );
  }

  // Independent-checker confirmation (empty body). 409 CONDITIONS_NOT_SATISFIED
  // with blocking ids, or 409 SAME_ACTOR_FORBIDDEN when the confirmer verified.
  async confirm(caseId: string): Promise<ConditionConfirmation> {
    return parseConditionConfirmation(
      await this.request(
        `/api/v1/cases/${encodeURIComponent(caseId)}/conditions/confirm`,
        { method: "POST", body: "{}" },
      ),
    );
  }

  private async request(path: string, init: RequestInit = {}): Promise<unknown> {
    const headers = new Headers(init.headers);
    headers.set("Accept", "application/json");
    if (init.body !== undefined) headers.set("Content-Type", "application/json");
    if (isMutation(init.method)) {
      const csrfToken = this.csrfTokenProvider();
      if (csrfToken) headers.set(CSRF_HEADER_NAME, csrfToken);
    }

    const response = await this.fetcher(`${this.baseUrl}${path}`, {
      ...init,
      headers,
      credentials: "include",
      cache: "no-store",
    });
    const body = await parseJson(response);
    if (!response.ok) {
      const apiError = parseApiError(body);
      throw new ApiClientError(
        response.status,
        apiError?.code ?? "REQUEST_FAILED",
        apiError?.messageVi || "Yêu cầu không thành công.",
        apiError?.retryable ?? response.status >= 500,
      );
    }
    return body;
  }
}

// Drops undefined / empty-array optional fields so the BFF's exact-keys check
// never sees a key the officer did not fill in.
function compact(input: object): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(input)) {
    if (value === undefined) continue;
    if (Array.isArray(value) && value.length === 0) continue;
    out[key] = value;
  }
  return out;
}

function isMutation(method: string | undefined): boolean {
  return method !== undefined && !["GET", "HEAD"].includes(method.toUpperCase());
}

export const conditionsApi = new ConditionsApiClient();

const GENERIC_MESSAGES = new Set(["Không thể hoàn tất yêu cầu.", "Yêu cầu không thành công."]);

export function getConditionError(error: unknown): string {
  if (error instanceof ApiClientError) {
    if (error.message && !GENERIC_MESSAGES.has(error.message)) {
      return error.message;
    }
    switch (error.status) {
      case 401:
        return "Phiên làm việc đã hết hạn. Vui lòng đăng nhập lại.";
      case 403:
        return "Bạn không có vai trò được yêu cầu cho thao tác điều kiện giải ngân.";
      case 404:
        return "Không tìm thấy hồ sơ hoặc điều kiện giải ngân. Vui lòng tải lại.";
      case 409:
        return "Không thể hoàn tất: trạng thái đã thay đổi hoặc điều kiện chưa đủ. Vui lòng tải lại.";
      case 422:
        return "Dữ liệu chưa hợp lệ. Vui lòng kiểm tra và thử lại.";
      case 503:
        return "Dịch vụ điều kiện giải ngân chưa sẵn sàng. Vui lòng thử lại sau ít phút.";
    }
  }
  return "Không thể hoàn tất yêu cầu. Vui lòng thử lại.";
}

// --- Display labels ---

export const GATE_STATUS_LABELS: Record<GateStatus, string> = {
  OPEN: "Đang chờ",
  SATISFIED: "Đạt",
};

export const CONDITION_STATUS_LABELS: Record<ConditionStatus, string> = {
  PENDING: "Chờ xử lý",
  EVIDENCE_SUBMITTED: "Đã nộp bằng chứng",
  VERIFIED: "Đã xác minh",
  FAILED: "Không đạt",
  WAIVER_REQUESTED: "Đề nghị miễn trừ",
  WAIVED_BY_HUMAN: "Đã miễn trừ (thẩm quyền)",
  SUPERSEDED: "Đã thay thế",
  NOT_APPLICABLE_BY_HUMAN: "Không áp dụng (thẩm quyền)",
};

// Precise per-target action labels (never a generic verb).
export const CONDITION_TRANSITION_LABELS: Record<ConditionStatus, string> = {
  PENDING: "Đưa về chờ xử lý",
  EVIDENCE_SUBMITTED: "Ghi nhận đã nộp bằng chứng",
  VERIFIED: "Xác minh điều kiện giải ngân",
  FAILED: "Ghi nhận không đạt",
  WAIVER_REQUESTED: "Đề nghị miễn trừ điều kiện",
  WAIVED_BY_HUMAN: "Miễn trừ điều kiện (ghi thẩm quyền)",
  SUPERSEDED: "Đánh dấu điều kiện đã thay thế",
  NOT_APPLICABLE_BY_HUMAN: "Xác định không áp dụng (ghi thẩm quyền)",
};

export const UNSUPPORTED_ENUM_LABEL = "Trạng thái chưa được hỗ trợ";

export function labelOrUnsupported<K extends string>(
  map: Record<K, string>,
  key: string,
): string {
  return (map as Record<string, string>)[key] ?? UNSUPPORTED_ENUM_LABEL;
}

export function shortId(value: string | null | undefined): string {
  if (!value) return "—";
  return value.length > 12 ? `${value.slice(0, 8)}…` : value;
}

export function formatDateTime(iso: string): string {
  const date = new Date(iso);
  if (Number.isNaN(date.getTime())) return iso || "—";
  return date.toLocaleString("vi-VN");
}

export function formatDate(iso: string | null): string {
  if (!iso) return "—";
  const date = new Date(iso);
  if (Number.isNaN(date.getTime())) return iso;
  return date.toLocaleDateString("vi-VN");
}

File: services/api/src/creditops/api/conditions.py
"""Disbursement ConditionLedger API: human-only condition lifecycle + the gate.

Master design section 5 giai đoạn 10.  Independent credit operations verify the
signed contract, perfected security, customer own-funds participation, purpose
documents, licences and other bound conditions BEFORE disbursement.  Four
surfaces, all case-scoped and fail-closed (an unassigned actor gets the same
indistinguishable 404 as a missing case):

- POST ``/conditions`` -- the ``OPS_OFFICER`` opens ONE condition.  A condition
  may only be opened once a PERMITTING human credit decision exists for the
  current case version (an approval: ``APPROVED_AS_PROPOSED`` /
  ``APPROVED_WITH_CONDITIONS``), loaded through the credit-decision repository.
  The condition binds that decision as its source.
- GET ``/conditions`` -- any case participant reads the ledger for the current
  version.
- POST ``/conditions/{id}/transition`` -- move ONE condition along a VALIDATED
  edge (``domain/conditions.py::ALLOWED_TRANSITIONS``).  AUTHORITY SPLIT
  (PROPOSED synthetic; no official SHB matrix): ``VERIFIED`` and the human
  ``WAIVED_BY_HUMAN`` / ``NOT_APPLICABLE_BY_HUMAN`` rulings require the
  independent ``OPS_CHECKER`` role; the waiver / not-applicable rulings
  additionally require an explicit authority rationale.  Every other move is an
  ordinary ``OPS_OFFICER`` workflow step.
- POST ``/conditions/confirm`` -- the independent ``OPS_CHECKER`` confirms the
  ledger.  It fails closed unless ``derive_conditions_confirmable`` is True (409
  ``CONDITIONS_NOT_SATISFIED`` listing the blocking ids -- an EMPTY ledger is
  never confirmable) AND the confirming actor is DIFFERENT from every actor who
  drove a ``VERIFIED`` transition (409 ``SAME_ACTOR_FORBIDDEN`` -- separation of
  duty).  On success it satisfies ``HG_DISBURSEMENT_CONDITIONS_CONFIRMED``
  through the orchestration repository (exactly as ``api/financing.py`` records
  its gate -- the gate-writing authority stays out of the condition port),
  audits, and re-ticks the orchestrator.

WAIVER is HUMAN-only with an authority record: a waiver is recorded as a status
transition to ``WAIVED_BY_HUMAN`` by the ``OPS_CHECKER`` role with a mandatory
rationale, captured on the append-only status-event trail.  No agent path exists
to any surface here.

This module is exported as ``router`` and is NOT registered in ``main.py`` here;
production wiring is a separate change (tests include the router directly).

All customer data in this project is synthetic and created solely for
demonstration.
"""

from __future__ import annotations

import logging
from datetime import date, datetime
from typing import Annotated, Any, cast
from uuid import UUID, uuid4

from fastapi import APIRouter, Depends, Request, Response
from pydantic import BaseModel, ConfigDict, Field, ValidationError

from creditops.api.auth import require_actor
from creditops.api.errors import ApiException
from creditops.application.orchestration.kickoff import KickoffOrchestration
from creditops.application.orchestration.roles import (
    CASE_PARTICIPANT_ROLES,
    OPS_OFFICER_ROLE,
)
from creditops.application.ports.conditions import (
    ConditionLedgerRepository,
    ConditionNotFound,
    ForbiddenConditionTransition,
    RecordedCondition,
)
from creditops.application.ports.credit_decisions import CreditDecisionRepository
from creditops.application.ports.orchestration import (
    OrchestrationAuditEvent,
    OrchestrationRepository,
)
from creditops.application.ports.repositories import CaseRecord
from creditops.application.unit_of_work import ActorContext
from creditops.application.use_cases.dispatch_outbox import DispatchOutbox
from creditops.domain.conditions import (
    CHECKER_AUTHORITY_TARGETS,
    CONFIRMABLE_STATUSES,
    RATIONALE_REQUIRED_TARGETS,
    ConditionStatus,
    DisbursementCondition,
    derive_conditions_confirmable,
    is_transition_allowed,
)
from creditops.domain.credit_decisions import APPROVAL_DECISIONS
from creditops.domain.orchestration import GateStatus, GateType
from creditops.observability import log_event

router = APIRouter(prefix="/api/v1/cases/{case_id}/conditions", tags=["conditions"])

_logger = logging.getLogger(__name__)

#: PROPOSED synthetic JWT authority role for the INDEPENDENT credit-operations
#: checker (design giai đoạn 10: "do ops checker độc lập thực hiện").  No
#: official SHB role exists; this dedicated role is the API-layer authority for
#: verification, human waiver / not-applicable rulings, and the final
#: confirmation, alongside a case assignment for row access.
OPS_CHECKER_ROLE = "OPS_CHECKER"

#: Roles allowed to READ the ledger: any case participant plus the ops checker.
_READ_ROLES = CASE_PARTICIPANT_ROLES | {OPS_CHECKER_ROLE}

#: Decision outcomes that PERMIT opening disbursement conditions (an approval).
_PERMITTING_DECISIONS: frozenset[str] = frozenset(
    decision.value for decision in APPROVAL_DECISIONS
)

#: PROPOSED synthetic disposition-reference prefix bound to the confirmed case
#: version (no official SHB mapping).
_DISPOSITION_REF_PREFIX = "disbursement-conditions"


class CreateConditionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    condition_text_vi: str = Field(
        alias="conditionText", min_length=1, max_length=4000
    )
    owner_vi: str | None = Field(default=None, alias="owner", min_length=1, max_length=400)
    due_date: date | None = Field(default=None, alias="dueDate")


class TransitionConditionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    to_status: str = Field(alias="toStatus", min_length=1, max_length=64)
    rationale_vi: str | None = Field(
        default=None, alias="rationale", min_length=1, max_length=4000
    )
    evidence_refs: tuple[str, ...] | None = Field(default=None, alias="evidenceRefs")


class ConditionResponse(BaseModel):
    model_config = ConfigDict(frozen=True, populate_by_name=True)

    id: UUID
    case_id: UUID = Field(serialization_alias="caseId")
    case_version: int = Field(serialization_alias="caseVersion")
    decision_id: UUID = Field(serialization_alias="decisionId")
    condition_text_vi: str = Field(serialization_alias="conditionText")
    owner_vi: str | None = Field(serialization_alias="owner")
    due_date: date | None = Field(serialization_alias="dueDate")
    status: str
    evidence_refs: list[str] = Field(serialization_alias="evidenceRefs")
    created_at: datetime = Field(serialization_alias="createdAt")


class ConditionsResponse(BaseModel):
    model_config = ConfigDict(frozen=True, populate_by_name=True)

    conditions: list[ConditionResponse]
    case_version: int = Field(serialization_alias="caseVersion")
    confirmable: bool


class ConfirmationResponse(BaseModel):
    model_config = ConfigDict(frozen=True, populate_by_name=True)

    gate_type: str = Field(serialization_alias="gateType")
    status: str
    case_version: int = Field(serialization_alias="caseVersion")
    disposition_ref: str = Field(serialization_alias="dispositionRef")


Actor = Annotated[ActorContext, Depends(require_actor)]


def _require_role(actor: ActorContext, role: str) -> None:
    if role not in actor.roles:
        raise ApiException(
            status_code=403,
            code="INSUFFICIENT_ROLE",
            message_vi="Bạn không có vai trò được yêu cầu cho thao tác này.",
        )


def _require_reader(actor: ActorContext) -> None:
    if not (_READ_ROLES & actor.roles):
        raise ApiException(
            status_code=403,
            code="INSUFFICIENT_ROLE",
            message_vi="Bạn không có vai trò tham gia hồ sơ được yêu cầu.",
        )


def _repository(request: Request) -> ConditionLedgerRepository:
    repository = getattr(request.app.state, "condition_ledger_repository", None)
    if repository is None:
        raise ApiException(
            status_code=503,
            code="CONDITION_LEDGER_SERVICE_UNAVAILABLE",
            message_vi="Dịch vụ điều kiện giải ngân chưa sẵn sàng.",
            retryable=True,
        )
    return cast(ConditionLedgerRepository, repository)


def _credit_decision_repository(request: Request) -> CreditDecisionRepository:
    repository = getattr(request.app.state, "credit_decision_repository", None)
    if repository is None:
        raise ApiException(
            status_code=503,
            code="CREDIT_DECISION_SERVICE_UNAVAILABLE",
            message_vi="Dịch vụ quyết định tín dụng chưa sẵn sàng.",
            retryable=True,
        )
    return cast(CreditDecisionRepository, repository)


def _orchestration_repository(request: Request) -> OrchestrationRepository | None:
    repository = getattr(request.app.state, "orchestration_repository", None)
    return cast("OrchestrationRepository | None", repository)


async def _assert_case_access(
    request: Request, actor: ActorContext, case_id: UUID
) -> CaseRecord:
    """Return the assigned case record, or fail closed with an indistinguishable
    404 for an unassigned actor (assignment membership is never disclosed)."""

    uow_factory = getattr(request.app.state, "uow_factory", None)
    if uow_factory is None:
        raise ApiException(
            status_code=503,
            code="CASE_SERVICE_UNAVAILABLE",
            message_vi="Dịch vụ hồ sơ chưa sẵn sàng.",
            retryable=True,
        )
    async with uow_factory(actor) as uow:
        record = await uow.cases.get_assigned(case_id, actor.actor_id)
    if record is None:
        raise ApiException(
            status_code=404,
            code="CASE_NOT_ACCESSIBLE",
            message_vi="Không tìm thấy hồ sơ hoặc bạn không có quyền truy cập.",
        )
    return cast(CaseRecord, record)


def _condition_response(condition: RecordedCondition) -> ConditionResponse:
    return ConditionResponse(
        id=condition.id,
        case_id=condition.case_id,
        case_version=condition.case_version,
        decision_id=condition.decision_id,
        condition_text_vi=condition.condition_text_vi,
        owner_vi=condition.owner_vi,
        due_date=condition.due_date,
        status=condition.status.value,
        evidence_refs=list(condition.evidence_refs),
        created_at=condition.created_at,
    )


@router.post("", response_model=ConditionResponse, status_code=201)
async def create_condition(
    case_id: UUID,
    body: CreateConditionRequest,
    actor: Actor,
    request: Request,
) -> ConditionResponse:
    """Open ONE disbursement condition bound to a permitting credit decision."""

    _require_role(actor, OPS_OFFICER_ROLE)
    record = await _assert_case_access(request, actor, case_id)

    decision = await _credit_decision_repository(request).load_decision(
        case_id, record.version
    )
    if decision is None or decision.decision not in _PERMITTING_DECISIONS:
        # No approval decision authorises a disbursement condition yet: fail
        # closed.  Conditions exist only downstream of a permitting decision.
        raise ApiException(
            status_code=409,
            code="CONDITIONS_REQUIRE_APPROVAL_DECISION",
            message_vi=(
                "Chưa có quyết định phê duyệt tín dụng cho phiên bản hồ sơ hiện "
                "tại để mở điều kiện giải ngân."
            ),
        )

    try:
        condition = DisbursementCondition(
            id=uuid4(),
            case_id=case_id,
            case_version=record.version,
            decision_id=decision.id,
            condition_text_vi=body.condition_text_vi,
            owner_vi=body.owner_vi,
            due_date=body.due_date,
            status=ConditionStatus.PENDING,
        )
    except (ValidationError, ValueError) as exc:
        raise ApiException(
            status_code=422,
            code="INVALID_CONDITION",
            message_vi="Điều kiện giải ngân không hợp lệ.",
        ) from exc

    created = await _repository(request).create_condition(
        condition=condition, actor_id=actor.actor_id, actor_role=OPS_OFFICER_ROLE
    )
    return _condition_response(created)


@router.get("", response_model=ConditionsResponse)
async def list_conditions(
    case_id: UUID, actor: Actor, request: Request
) -> ConditionsResponse:
    _require_reader(actor)
    record = await _assert_case_access(request, actor, case_id)
    conditions = await _repository(request).list_conditions(case_id, record.version)
    return ConditionsResponse(
        conditions=[_condition_response(c) for c in conditions],
        case_version=record.version,
        confirmable=derive_conditions_confirmable(c.status for c in conditions),
    )


@router.post("/{condition_id}/transition", response_model=ConditionResponse)
async def transition_condition(
    case_id: UUID,
    condition_id: UUID,
    body: TransitionConditionRequest,
    actor: Actor,
    request: Request,
) -> ConditionResponse:
    """Move ONE condition along a validated edge under the correct authority."""

    try:
        to_status = ConditionStatus(body.to_status)
    except ValueError as exc:
        raise ApiException(
            status_code=422,
            code="INVALID_CONDITION_STATUS",
            message_vi="Trạng thái điều kiện không hợp lệ.",
        ) from exc

    # The authority required depends on the target: VERIFIED and the human
    # waiver / not-applicable rulings are independent-checker acts; every other
    # move is an ordinary ops-officer step.
    required_role = (
        OPS_CHECKER_ROLE if to_status in CHECKER_AUTHORITY_TARGETS else OPS_OFFICER_ROLE
    )
    _require_role(actor, required_role)
    record = await _assert_case_access(request, actor, case_id)

    if to_status in RATIONALE_REQUIRED_TARGETS and not body.rationale_vi:
        # Waiver / not-applicable are human authority acts: the rationale IS the
        # authority record and is mandatory.
        raise ApiException(
            status_code=422,
            code="RATIONALE_REQUIRED",
            message_vi="Cần lý do (authority record) cho quyết định miễn trừ / không áp dụng.",
        )

    repository = _repository(request)
    current = await repository.load_condition(condition_id, case_id, record.version)
    if current is None:
        raise ApiException(
            status_code=404,
            code="CONDITION_NOT_FOUND",
            message_vi="Không tìm thấy điều kiện giải ngân trong hồ sơ này.",
        )
    if not is_transition_allowed(current.status, to_status):
        raise ApiException(
            status_code=422,
            code="FORBIDDEN_CONDITION_TRANSITION",
            message_vi="Chuyển trạng thái điều kiện không được phép.",
            details={"fromStatus": current.status.value, "toStatus": to_status.value},
        )

    try:
        updated = await repository.transition_condition(
            condition_id=condition_id,
            case_id=case_id,
            case_version=record.version,
            to_status=to_status,
            actor_id=actor.actor_id,
            actor_role=required_role,
            rationale_vi=body.rationale_vi,
            evidence_refs=body.evidence_refs,
        )
    except ConditionNotFound as exc:
        raise ApiException(
            status_code=404,
            code="CONDITION_NOT_FOUND",
            message_vi="Không tìm thấy điều kiện giải ngân trong hồ sơ này.",
        ) from exc
    except ForbiddenConditionTransition as exc:
        # Lost race: the condition moved between the pre-check and the write.
        raise ApiException(
            status_code=422,
            code="FORBIDDEN_CONDITION_TRANSITION",
            message_vi="Chuyển trạng thái điều kiện không được phép.",
        ) from exc
    return _condition_response(updated)


@router.post("/confirm", response_model=ConfirmationResponse)
async def confirm_conditions(
    case_id: UUID,
    actor: Actor,
    request: Request,
    response: Response,
) -> ConfirmationResponse:
    """Independent OPS-checker confirmation of the disbursement-condition ledger."""

    _require_role(actor, OPS_CHECKER_ROLE)
    record = await _assert_case_access(request, actor, case_id)
    repository = _repository(request)

    conditions = await repository.list_conditions(case_id, record.version)
    if not derive_conditions_confirmable(c.status for c in conditions):
        blocking = [
            str(c.id) for c in conditions if c.status not in CONFIRMABLE_STATUSES
        ]
        raise ApiException(
            status_code=409,
            code="CONDITIONS_NOT_SATISFIED",
            message_vi=(
                "Còn điều kiện giải ngân chưa được xác minh / miễn trừ / xác định "
                "không áp dụng; không thể xác nhận."
            ),
            details={"blockingConditionIds": blocking, "empty": not conditions},
        )

    # Separation of duty: the confirming checker must not be any actor who drove
    # a VERIFIED transition (the maker/checker cannot self-confirm their own
    # verification).  PROPOSED: exclusion is scoped to VERIFIED actors per the
    # design ("confirming actor must differ from every actor who VERIFIED").
    verifying_actors = await repository.list_verifying_actor_ids(
        case_id, record.version
    )
    if actor.actor_id in verifying_actors:
        raise ApiException(
            status_code=409,
            code="SAME_ACTOR_FORBIDDEN",
            message_vi=(
                "Người xác nhận độc lập phải khác với người đã xác minh điều kiện."
            ),
            details={"actorId": str(actor.actor_id)},
        )

    orchestration = _orchestration_repository(request)
    if orchestration is None:
        raise ApiException(
            status_code=503,
            code="ORCHESTRATION_SERVICE_UNAVAILABLE",
            message_vi="Dịch vụ điều phối chưa sẵn sàng.",
            retryable=True,
        )

    disposition_ref = f"{_DISPOSITION_REF_PREFIX}:{record.version}"
    await orchestration.ensure_gate(
        case_id=case_id,
        case_version=record.version,
        gate_type=GateType.HG_DISBURSEMENT_CONDITIONS_CONFIRMED,
        status=GateStatus.SATISFIED,
        satisfied_by_actor_id=actor.actor_id,
        disposition_ref=disposition_ref,
    )
    await repository.append_audit(
        OrchestrationAuditEvent(
            case_id=case_id,
            case_version=record.version,
            event_type="DISBURSEMENT_CONDITIONS_CONFIRMED",
            execution_id=uuid4(),
            artifact_type="DISBURSEMENT_CONDITION_LEDGER",
            artifact_id=case_id,
            event_data={
                "actorId": str(actor.actor_id),
                "caseVersion": record.version,
                "conditionCount": len(conditions),
            },
        )
    )
    await _retick_orchestration(
        request, orchestration, case_id=case_id, trigger_ref=f"HG_COND:{record.version}"
    )
    response.status_code = 200
    return ConfirmationResponse(
        gate_type=GateType.HG_DISBURSEMENT_CONDITIONS_CONFIRMED.value,
        status=GateStatus.SATISFIED.value,
        case_version=record.version,
        disposition_ref=disposition_ref,
    )


async def _retick_orchestration(
    request: Request,
    orchestration_repository: Any,
    *,
    case_id: UUID,
    trigger_ref: str,
) -> None:
    """Self-fire an idempotent orchestration tick after a gate satisfaction.

    The plan task + outbox event commit durably; the queue publish is
    best-effort (the recovery dispatch picks up anything left).  A tick failure
    must never fail the human's already-recorded confirmation, but it is logged,
    never silent.
    """

    try:
        result = await KickoffOrchestration(orchestration_repository).execute(
            case_id, trigger_ref=trigger_ref
        )
        queue = getattr(request.app.state, "agent_task_queue", None)
        if queue is not None:
            await DispatchOutbox(
                orchestration_repository,
                queue,
                worker_dispatcher=getattr(request.app.state, "worker_dispatcher", None),
            ).run()
        log_event(
            _logger,
            logging.INFO,
            "Orchestration retick after gate satisfaction",
            {
                "event": "orchestration_retick",
                "trigger": trigger_ref,
                "created": result.created,
            },
        )
    except Exception:
        log_event(
            _logger,
            logging.ERROR,
            "Orchestration retick failed; the confirmation is durable and the "
            "case can be advanced manually",
            {"event": "orchestration_retick_failed", "trigger": trigger_ref},
        )
